tracer/src/Datadog.Trace/ClrProfiler/AutoInstrumentation/IbmMq/IbmMqHelper.cs

```diff
@@ -0,0 +1,133 @@
+
+            var resourceName = $"Produce Topic {queue.Name}";
+
+            var span = scope.Span;
+            span.Type = SpanTypes.Queue;
+            span.ResourceName = resourceName;
+            span.SetTag(Tags.SpanKind, SpanKinds.Producer);
+
+            var adapter = new IbmMqHeadersAdapter(message);
+            SpanContextPropagator.Instance.Inject(span.Context, adapter);
+        }
+        catch (Exception ex)
+        {
+            Log.Error(ex, "Error creating or populating scope.");
+        }
+
+        return scope;
+    }
+
+    internal static Scope? CreateConsumerScope(
+        Tracer tracer,
+        IMqQueue queue,
+        IMqMessage message)
+    {
+        Scope? scope = null;
+
+        try
+        {
+            var settings = tracer.Settings;
+            if (!settings.IsIntegrationEnabled(IntegrationId.IbmMq))
+            {
+                return null;
+            }
+
+            var parent = tracer.ActiveScope?.Span;
+            var operationName = tracer.CurrentTraceSettings.Schema.Messaging.GetInboundOperationName(MessagingType);
+            if (parent is not null &&
+                parent.OperationName == operationName &&
+                parent.GetTag(Tags.InstrumentationName) != null)
+            {
+                return null;
+            }
+
+            SpanContext? propagatedContext = null;
+            PathwayContext? pathwayContext = null;
+
+            var adapter = new IbmMqHeadersAdapter(message);
+            try
+            {
+                propagatedContext = SpanContextPropagator.Instance.Extract(adapter);
+            }
+            catch (Exception ex)
+            {
+                Log.Error(ex, "Error extracting propagated headers from IbmMq message");
+            }
+
+            var dataStreams = tracer.TracerManager.DataStreamsManager;
+            if (dataStreams.IsEnabled)
+            {
+                try
+                {
+                    pathwayContext = dataStreams.ExtractPathwayContextAsBase64String(adapter);
+                }
+                catch (Exception ex)
+                {
+                    Log.Error(ex, "Error extracting PathwayContext from IbmMq message");
+                }
+            }
+
+            var serviceName = tracer.CurrentTraceSettings.Schema.Messaging.GetServiceName(MessagingType);
+            scope = tracer.StartActiveInternal(operationName, parent: propagatedContext, serviceName: serviceName, finishOnClose: true);
+            tracer.TracerManager.Telemetry.IntegrationGeneratedSpan(IntegrationId.IbmMq);
+
+            var resourceName = $"Consume Topic {(string.IsNullOrEmpty(queue.Name) ? "ibmmq" : queue.Name)}";
+
+            var span = scope.Span;
+            span.Type = SpanTypes.Queue;
+            span.ResourceName = resourceName;
+            span.SetTag(Tags.SpanKind, SpanKinds.Consumer);
+            span.SetTag(Tags.Measured, "1");
```

Comment:
Should we be adding this @lucaspimentel? 🤔 I don't think we should, but I get forget😅

Reply:
No, we don't need to explicitly add this tag.

Either this span's service name will be different than its parent's service name, so it will be marked as a top-level span, which are implicitly "measured," or a newer agent version is present to support peer service, and newer agent version automatically "measure" all spans where `span.kind` is `client`, `server`, `producer`, or `consumer`.

Reply:
```suggestion
```